src/probability_theory/martingale.lean

```diff
@@ -83,11 +112,149 @@ begin
   rw [pi.smul_apply, hx, pi.smul_apply, pi.smul_apply],
 end
 
+lemma supermartingale [preorder E] (hf : martingale f ℱ μ) : supermartingale f ℱ μ :=
+⟨hf.1, λ i j hij, (hf.2 i j hij).le, λ i, hf.integrable i⟩
+
+lemma submartingale [preorder E] (hf : martingale f ℱ μ) : submartingale f ℱ μ :=
+⟨hf.1, λ i j hij, (hf.2 i j hij).symm.le, λ i, hf.integrable i⟩
+
 end martingale
 
+lemma martingale_iff [partial_order E] : martingale f ℱ μ ↔
+  supermartingale f ℱ μ ∧ submartingale f ℱ μ :=
+⟨λ hf, ⟨hf.supermartingale, hf.submartingale⟩,
+ λ ⟨hf₁, hf₂⟩, ⟨hf₁.1, λ i j hij, (hf₁.2.1 i j hij).antisymm (hf₂.2.1 i j hij)⟩⟩
+
 lemma martingale_condexp (f : α → E) (ℱ : filtration ι m0) (μ : measure α)
   [sigma_finite_filtration μ ℱ] :
   martingale (λ i, μ[f | ℱ i, ℱ.le i]) ℱ μ :=
 ⟨λ i, measurable_condexp, λ i j hij, condexp_condexp_of_le (ℱ.mono hij) _⟩
 
+namespace supermartingale
+
+lemma adapted [has_le E] (hf : supermartingale f ℱ μ) : adapted ℱ f := hf.1
+
+lemma measurable [has_le E] (hf : supermartingale f ℱ μ) (i : ι) : measurable[ℱ i] (f i) :=
+hf.adapted i
+
+lemma integrable [has_le E] (hf : supermartingale f ℱ μ) (i : ι) : integrable (f i) μ := hf.2.2 i
+
+lemma condexp_ae_eq [has_le E] (hf : supermartingale f ℱ μ) {i j : ι} (hij : i ≤ j) :
+  μ[f j | ℱ i, ℱ.le i] ≤ᵐ[μ] f i :=
+hf.2.1 i j hij
+
+lemma set_integral_le {f : ι → α → ℝ} (hf : supermartingale f ℱ μ)
+  {i j : ι} (hij : i ≤ j) {s : set α} (hs : measurable_set[ℱ i] s) :
+  ∫ x in s, f j x ∂μ ≤ ∫ x in s, f i x ∂μ :=
+begin
+  rw ← @set_integral_condexp _ _ _ _ _ _ _ _ (ℱ i) m0 _ (ℱ.le i) _ _ _ (hf.integrable j) hs,
+  refine set_integral_mono_ae integrable_condexp.integrable_on (hf.integrable i).integrable_on _,
+  filter_upwards [hf.2.1 i j hij],
+  intros _ heq,
+  exact heq,
+end
+
+lemma add [preorder E] [covariant_class E E (+) (≤)]
+  (hf : supermartingale f ℱ μ) (hg : supermartingale g ℱ μ) : supermartingale (f + g) ℱ μ :=
+begin
+  refine ⟨hf.1.add hg.1, λ i j hij, _, λ i, (hf.2.2 i).add (hg.2.2 i)⟩,
+  refine (condexp_add (hf.integrable j) (hg.integrable j)).le.trans _,
+  filter_upwards [hf.2.1 i j hij, hg.2.1 i j hij],
+  intros,
+  refine add_le_add _ _; assumption,
+end
+
+lemma add_martingale [preorder E] [covariant_class E E (+) (≤)]
+  (hf : supermartingale f ℱ μ) (hg : martingale g ℱ μ) : supermartingale (f + g) ℱ μ :=
+hf.add hg.supermartingale
+
+lemma neg [preorder E] [covariant_class E E (+) (≤)]
+  (hf : supermartingale f ℱ μ) : submartingale (-f) ℱ μ :=
+begin
+  refine ⟨hf.1.neg, λ i j hij, _, λ i, (hf.2.2 i).neg⟩,
+  refine eventually_le.trans _ (condexp_neg (f j)).symm.le,
+  filter_upwards [hf.2.1 i j hij],
+  intros _ hle,
+  simpa,
+end
+
+-- We need a notion of "ordered_normed_group" for the following
```

Comment:
there is `normed_lattice_add_comm_group`, which is stronger than what you strictly need (since it has a lattice while a preorder would be enough).

Reply:
I can't seem to get it to work. I'm stuck on 
```lean
lemma smul_nonneg [normed_lattice_add_comm_group E]
  {c : ℝ} (hc : 0 ≤ c) (hf : supermartingale f ℱ μ) :
  supermartingale (c • f) ℱ μ :=
begin
  refine ⟨hf.1.smul c, λ i j hij, _, λ i, (hf.2.2 i).smul c⟩,
  refine (condexp_smul c (f j)).le.trans _,
  filter_upwards [hf.2.1 i j hij],
  intros _ hle,
  simp,
  refine smul_le_smul_of_nonneg  hle hc, 
  -- failed to synthesize type class instance for smul_with_zero ℝ E
end
```
Do you think we are missing a lemma?

Reply:
I can't really tell since this is not a mwe and I don't see all typeclasses you have put on `E`, but I can guess possible problems: if you declared `[normed_group E]` somewhere and add `[normed_lattice_add_comm_group E]` then you have two normed group structures, which is not good. Also, do you have `[normed_space R E]`?

Reply:
I think this is a representative MWE:
```lean
import measure_theory.function.conditional_expectation

open topological_space filter
open_locale nnreal ennreal measure_theory

variables {F : Type*} [measurable_space F] [normed_lattice_add_comm_group F]
  [normed_space ℝ F] [complete_space F] [borel_space F] [second_countable_topology F]

example (c : ℝ) (hc : 0 ≤ c) (f g : F) (h : c • f ≤ c • g) : f ≤ g := 
begin
  sorry
end
```
I don't think I have multiple normed group structures in this example. I think my main problem is I can't use `smul_le_smul_of_nonneg` since we don't have `smul_with_zero`.

Reply:
does it work if you add `[ordered_smul ℝ F]` to the list of hypotheses?

Reply:
Great that works!